File: js/src/orchestration/intent.ts
/**
 * Generic intent types and resolvers for the orchestration layer.
 */

export type GenericIntentObject =
  | "greeting"
  | "query"
  | "action"
  | "confirmation"
  | "unknown";

export interface OrchestratorIntent<T extends string = GenericIntentObject> {
  object: T;
  confidence: number;
  requiresConfirmation?: boolean;
}

export interface IntentResolveInput {
  message: string;
  history?: unknown[];
  sessionState?: unknown;
}

export interface BaseIntentResolver<T extends string = GenericIntentObject> {
  resolve(input: IntentResolveInput): Promise<OrchestratorIntent<T>>;
}

export interface DeterministicPattern<T extends string = GenericIntentObject> {
  pattern: RegExp;
  intent: OrchestratorIntent<T>;
}

/**
 * DeterministicIntentResolver — fast pattern matching, no LLM calls.
 * Returns first matching pattern or unknown with confidence 0.
 */
export class DeterministicIntentResolver<T extends string = GenericIntentObject>
  implements BaseIntentResolver<T> {
  private patterns: DeterministicPattern<T>[];

  constructor(patterns: DeterministicPattern<T>[]) {
    this.patterns = patterns;
  }

  async resolve(input: IntentResolveInput): Promise<OrchestratorIntent<T>> {
    for (const entry of this.patterns) {
      if (entry.pattern.test(input.message)) {
        return entry.intent;
      }
    }
    return { object: "unknown" as T, confidence: 0 };
  }
}

/**
 * LLMFallbackIntentResolver — tries deterministic first, falls back to LLM classifier.
 */
export class LLMFallbackIntentResolver<T extends string = GenericIntentObject>
  implements BaseIntentResolver<T> {
  private primary: BaseIntentResolver<T>;
  private fallbackAgent: { chat(message: string): Promise<{ content: string }> };
  private parseResponse?: (content: string) => OrchestratorIntent<T>;

  constructor(config: {
    primary: BaseIntentResolver<T>;
    fallbackAgent: { chat(message: string): Promise<{ content: string }> };
    parseResponse?: (content: string) => OrchestratorIntent<T>;
  }) {
    this.primary = config.primary;
    this.fallbackAgent = config.fallbackAgent;
    this.parseResponse = config.parseResponse;
  }

  async resolve(input: IntentResolveInput): Promise<OrchestratorIntent<T>> {
    const primaryResult = await this.primary.resolve(input);

    if (primaryResult.object !== "unknown") {
      return primaryResult;
    }

    // Fallback to LLM
    const llmResponse = await this.fallbackAgent.chat(
      `Classify the intent of this message: "${input.message}". Return a JSON object with "object" (string) and "confidence" (number 0-1).`,
    );

    if (this.parseResponse) {
      return this.parseResponse(llmResponse.content);
    }

    // Default parsing
    try {
      const parsed = JSON.parse(llmResponse.content);
      return {
        object: (parsed.object ?? "unknown") as T,
        confidence: parsed.confidence ?? 0.5,
      };
    } catch {
      return { object: "unknown" as T, confidence: 0 };
    }
  }
}

File: naru_agent/orchestration/intent.py
"""Intent types and resolvers for the orchestration layer."""

from __future__ import annotations

import json
import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Generic, TypeVar

logger = logging.getLogger(__name__)

GenericIntentObject = str  # "greeting" | "query" | "action" | "confirmation" | "unknown"

T = TypeVar("T", bound=str)


@dataclass
class OrchestratorIntent(Generic[T]):
    """Resolved intent from an intent resolver."""

    object: T  # noqa: A003
    confidence: float
    requires_confirmation: bool = False


@dataclass
class IntentResolveInput:
    """Input to an intent resolver."""

    message: str
    history: list[Any] = field(default_factory=list)
    session_state: Any = None


class BaseIntentResolver(ABC, Generic[T]):
    """Abstract base for intent resolvers."""

    @abstractmethod
    def resolve(self, input: IntentResolveInput) -> OrchestratorIntent[T]:  # noqa: A002
        ...


@dataclass
class DeterministicPattern(Generic[T]):
    """A regex pattern mapped to a fixed intent."""

    pattern: Any  # re.Pattern[str]
    intent: OrchestratorIntent[T]


class DeterministicIntentResolver(BaseIntentResolver[T]):
    """Fast pattern matching, no LLM calls.

    Returns the first matching pattern's intent, or an 'unknown' intent with
    confidence 0 when no pattern matches.
    """

    def __init__(self, patterns: list[DeterministicPattern[T]]) -> None:
        self._patterns = patterns

    def resolve(self, input: IntentResolveInput) -> OrchestratorIntent[T]:  # noqa: A002
        for entry in self._patterns:
            if entry.pattern.search(input.message):
                return entry.intent
        return OrchestratorIntent(object="unknown", confidence=0.0)  # type: ignore[arg-type]


class LLMFallbackIntentResolver(BaseIntentResolver[T]):
    """Tries deterministic first, falls back to an LLM classifier.

    The fallback_agent only needs a ``chat(message: str)`` method that returns
    an object with a ``content`` attribute (str).  NaruAgent satisfies this.
    """

    def __init__(
        self,
        primary: BaseIntentResolver[T],
        fallback_agent: Any,
        parse_response: Any = None,
    ) -> None:
        self._primary = primary
        self._fallback_agent = fallback_agent
        self._parse_response = parse_response

    def resolve(self, input: IntentResolveInput) -> OrchestratorIntent[T]:  # noqa: A002
        primary_result = self._primary.resolve(input)

        if primary_result.object != "unknown":
            return primary_result

        # Fallback to LLM
        llm_response = self._fallback_agent.chat(
            f'Classify the intent of this message: "{input.message}". '
            'Return a JSON object with "object" (string) and "confidence" (number 0-1).'
        )

        if self._parse_response is not None:
            return self._parse_response(llm_response.content)

        # Default parsing
        try:
            parsed = json.loads(llm_response.content)
            return OrchestratorIntent(
                object=parsed.get("object", "unknown"),
                confidence=float(parsed.get("confidence", 0.5)),
            )
        except Exception:  # noqa: BLE001
            return OrchestratorIntent(object="unknown", confidence=0.0)  # type: ignore[arg-type]
